compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -690,6 +690,29 @@ trait Implicits { self: Typer =>
       }
     }
 
+    /** Creates a tree that will produce a ValueOf instance for the requested type.
+      * An EmptyTree is returned if materialization fails.
+      */
+    def synthesizedValueOf(formal: Type)(implicit ctx: Context): Tree = {
+      def success(t: Tree) = New(defn.ValueOfClass.typeRef.appliedTo(t.tpe), t :: Nil).withPos(pos)
+
+      formal.argTypes match {
+        case arg :: Nil =>
+          fullyDefinedType(arg.dealias, "ValueOf argument", pos) match {
+            case ConstantType(c: Constant) =>
+              success(Literal(c))
+            case TypeRef(_, sym) if sym == defn.UnitClass =>
+              success(Literal(Constant(())))
```

Comment:
Isn't `Unit` handled by the `ContantType` case?

Reply:
No it isn't.

Reply:
Perhaps it should be, but currently `Unit` isn't classified as a singleton type in either Scala 2 or Dotty.

Reply:
It should certainly be, at least conceptually. Isn't `Unit` the singleton type _par excellence_?